cmd/controller-manager/app/controller-manager.go

```diff
@@ -104,6 +107,36 @@ func Run(opts *options.Options) error {
 		glog.Info("Federation will target all namespaces")
 	}
 
+	if !opts.LeaderElection.LeaderElect {
+		// Leader election is disabled, so run inline until done.
+		startControllers(opts, stopChan)
+		<-stopChan
+		return errors.New("finished without leader election")
+	}
+
+	lec, err := leaderelection.LeaderElector(opts, startControllers)
+	if err != nil {
+		panic(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go func() {
+		select {
+		case <-stopChan:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+
+	lec.Run(ctx)
```

Comment:
Will this loop forever?
A brief look at the package code and its relevant use in k8s seems to indicate that it does not.
I did not find `Run` calling the handler `OnStartedLeading` whenever it gains leadership in a loop.
k8s ctrl-mgr seem to [implement usage](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L252) such that the process exits with error if it loses leadership; the pod should ideally re-start the process and while starting it would wait to gain leadership if there already is an existing leader.
Should we also be doing the same?

Reply:
Here's the implementation of `Run`:

```golang
func (le *LeaderElector) Run(ctx context.Context) {                                                                          
    defer func() {                                                                                                           
        runtime.HandleCrash()                                                                                                
        le.config.Callbacks.OnStoppedLeading()                                                                               
    }()                                                                                                                      
    if !le.acquire(ctx) {                                                                                                    
        return // ctx signalled done                                                                                         
    }                                                                                                                        
    ctx, cancel := context.WithCancel(ctx)                                                                                   
    defer cancel()                                                                                                           
    go le.config.Callbacks.OnStartedLeading(ctx)                                                                             
    le.renew(ctx)                                                                                                            
}    
```

Check the `acquire` method's use of `JitterUntil`.  Once acquired, `OnStartedLeading` gets called and then `renew` gets called, which will loop until no longer the leader and then exit.  An exit will prompt the container to be restarted, and it will loop in `acquire` until it is elected the master.  I think this looks solid.

Reply:
Yeah, I did miss the simple fact that the container/pod default restart policy is `restartPolicy: Always` and would restart the process on a successful exit  also.